src/Components/Server/src/Circuits/CircuitRegistry.cs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal class CircuitRegistry
+    {
+        private readonly ComponentsServerOptions _options;
+        private readonly ILogger _logger;
+        private readonly PostEvictionCallbackRegistration _postEvictionCallback;
+
+        public CircuitRegistry(
+            IOptions<ComponentsServerOptions> options,
+            ILogger<CircuitRegistry> logger)
+        {
+            _options = options.Value;
+            _logger = logger;
+
+            ActiveCircuits = new ConcurrentDictionary<string, CircuitHost>(StringComparer.Ordinal);
+
+            InactiveCircuits = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _options.MaxRetainedDisconnectedCircuits,
+            });
+
+            _postEvictionCallback = new PostEvictionCallbackRegistration
+            {
+                EvictionCallback = OnEntryEvicted,
+            };
+        }
+
+        internal ConcurrentDictionary<string, CircuitHost> ActiveCircuits { get; }
+
+        internal MemoryCache InactiveCircuits { get; }
+
+        public void Register(CircuitHost circuitHost)
+        {
+            ActiveCircuits.TryAdd(circuitHost.CircuitId, circuitHost);
+        }
+
+        public void Deactivate(CircuitHost circuitHost, string connectionId)
+        {
+            if (!ActiveCircuits.TryGetValue(circuitHost.CircuitId, out circuitHost))
```

Comment:
I suppose what I expected to see here was a big explanation in a comment of what the states are, and how we manage the transitions between them. I'm trying to put the pieces together while reading this over, but I think it needs a whole lot more explanation - especially details about how the state transitions work.

Trying to manage adding and removing state between two collections is just fundamentally hard without locking.